File: Array/wuchao/stack/6-17栈实现队列/MyQueue.c
#include <stdbool.h>
#include <stdlib.h>
//
// Created by 吴超 on 2020/6/17.
//

typedef struct {
    int value;
    struct Node* next;
} Node;
Node* createNode(int x);
typedef struct {
    Node* head;
} MyStack;
MyStack* myStackCreate();
void myStackPush(MyStack* obj, int x);
int myStackPop(MyStack* obj);
int myStackTop(MyStack* obj);
bool myStackEmpty(MyStack* obj);

Node* createNode(int x){
    Node* node = malloc(sizeof(Node));
    node->value = x;
    node->next = NULL;
    return node;
}

MyStack* myStackCreate(){
    MyStack* myStack = malloc(sizeof(MyStack));
    myStack->head = createNode(-1);
    return myStack;
}

bool myStackEmpty(MyStack* obj){
    return obj->head->next == NULL;
}

void myStackPush(MyStack* obj, int x){
    Node* node = createNode(x);
    node->next = obj->head;
    obj->head = node;
}

int myStackPop(MyStack* obj){
    if(myStackEmpty(obj)){
        return -1;
    }
    int top = myStackTop(obj);
    obj->head = obj->head->next;
    return top;
}

int myStackTop(MyStack* obj){
    return obj->head->value;
}

typedef struct {
    MyStack * temp;
    MyStack * stack;
} MyQueue;
bool myQueueEmpty(MyQueue* obj);

/** Initialize your data structure here. */

MyQueue* myQueueCreate() {
    MyQueue* myQueue = malloc(sizeof(MyQueue));
    myQueue->temp = myStackCreate();
    myQueue->stack = myStackCreate();
    return myQueue;
}

/** Push element x to the back of queue. */
void myQueuePush(MyQueue* obj, int x) {
    while(!myStackEmpty(obj->stack)){
        myStackPush(obj->temp,myStackPop(obj->stack));
    }
    myStackPush(obj->stack,x);
    while(!myStackEmpty(obj->temp)){
        myStackPush(obj->stack,myStackPop(obj->temp));
    }
}

/** Removes the element from in front of queue and returns that element. */
int myQueuePop(MyQueue* obj) {
    return myStackPop(obj->stack);
}

/** Get the front element. */
int myQueuePeek(MyQueue* obj) {
    return myStackTop(obj->stack);
}

/** Returns whether the queue is empty. */
bool myQueueEmpty(MyQueue* obj) {
    return myStackEmpty(obj->stack);
}

void myQueueFree(MyQueue* obj) {
    free(obj->temp);
    free(obj->stack);
    free(obj);
}

File: Array/zhouhongkun/StackToQueue/src/com/chow/MyQueue.java
package com.chow;

import java.util.Stack;

public class MyQueue {

    Stack<Integer> stack1;
    Stack<Integer> stack2;

    /** Initialize your data structure here. */
    public MyQueue() {
        stack1 = new Stack<>();
        stack2 = new Stack<>();
    }

    /** Push element x to the back of queue. */
    public void push(int x) {
        stack1.push(x);
    }

    /** Removes the element from in front of queue and returns that element. */
    public int pop() {
        peek();
        return stack2.pop();
    }

    /** Get the front element. */
    public int peek() {
        if(stack2.isEmpty()){
            while (!stack1.isEmpty()){
                stack2.push(stack1.pop());
            }
            return stack2.peek();
        }
        return stack2.peek();
    }

    /** Returns whether the queue is empty. */
    public boolean empty() {
        return stack1.isEmpty() && stack2.isEmpty();
    }
}
